src/QueryEditor.tsx

```diff
@@ -1,47 +1,74 @@
 import React from 'react';
-import { css } from '@emotion/css';
 import { useAsync } from 'react-use';
-import { CoreApp, QueryEditorProps, SelectableValue } from '@grafana/data';
+import { QueryEditorProps, SelectableValue } from '@grafana/data';
 import { DataFrameDataSource } from './datasource';
-import { Column, DataframeQuery, QueryColumn } from './types';
-import { Button, InlineField, Input, MultiSelect } from '@grafana/ui';
+import { Column, DataframeQuery, isValidQuery, QueryColumn } from './types';
+import { InlineField, InlineSwitch, Input, MultiSelect, Select } from '@grafana/ui';
+import { decimationMethods, defaultDecimationMethod } from './constants';
 
 type Props = QueryEditorProps<DataFrameDataSource, DataframeQuery>;
 
-export const QueryEditor: React.FC<Props> = ({ app, query, datasource, onChange, onRunQuery }) => {
+export const QueryEditor: React.FC<Props> = ({ query, datasource, onChange, onRunQuery }) => {
   const tableMetadata = useAsync(async () => {
     return query.tableId ? await datasource.getTableMetadata(query.tableId) : null;
   }, [query.tableId]);
 
+  const runQueryIfValid = () => isValidQuery(query) && onRunQuery();
+
+  const handleQueryChange = (value: DataframeQuery, runQuery?: boolean) => {
```

Comment:
`runQuery?: boolean`
Is there a reason to make this optional over passing false in all the places where we don't want to run the query?

If not, I'd rather force callers to explicitly declare whether they want to run the query while handling the change or not. It makes it more immediately apparent at the call sites whether `handleQueryChange` is going to run the query. It also forces developers to make an active decision about it if we add more calls to `handleQueryChange` in the future.

Reply:
I feel ambivalent about this. I think patterns like this are common in TypeScript/JS because it works well with the implicit boolean conversions (undefined and false are treated the same). I like being able to ignore the parameter at the call site if it isn't needed.

> It also forces developers to make an active decision about it if we add more calls to handleQueryChange in the future.

However, this argument is compelling in this specific case. I think I'll make it explicit.